lib/client/extended_keyring.go

```diff
@@ -474,6 +491,98 @@ func callKeyExtension(agent agent.ExtendedAgent) (*profile.Profile, *ForwardedKe
 	return &profile, &forwardedKey, nil
 }
 
+// KeyExtensionResponse is a request object for the prompt-mfa-challenge@goteleport.com extension.
+type promptMFAChallengeRequest struct {
+	// ProxyAddr is the teleport proxy address that this challenge originated from.
+	ProxyAddr string
+	// MFAChallengeBlob is a json encoded *proto.MFAAuthenticateChallenge.
+	MFAChallengeBlob []byte
+	// MFAChallengeOptsBlob is a json encoded *proto.PromptMFAChallengeOpts.
+	MFAChallengeOptsBlob []byte
+}
+
+// KeyExtensionResponse is a response object for the prompt-mfa-challenge@goteleport.com extension.
+type promptMFAChallengeResponse struct {
+	// MFAChallengeResponseBlob is a json encoded *proto.MFAAuthenticateRespons.
+	MFAChallengeResponseBlob []byte
+}
+
+// promptMFAChallengeExtensionHandler returns an extensionHandler for the
+// prompt-mfa-challenge@goteleport.com extension.
+func promptMFAChallengeExtensionHandler(r *extendedKeyring) extensionHandler {
+	return func(contents []byte) ([]byte, error) {
+		var req promptMFAChallengeRequest
+		if err := ssh.Unmarshal(contents, &req); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		var challengeReq proto.MFAAuthenticateChallenge
+		if err := json.Unmarshal(req.MFAChallengeBlob, &challengeReq); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		var challengeOpts PromptMFAChallengeOpts
+		if err := json.Unmarshal(req.MFAChallengeOptsBlob, &challengeOpts); err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		mfaResp, err := promptMFAStandalone(context.Background(), &challengeReq, req.ProxyAddr, &challengeOpts)
```

Comment:
Maybe there's a way to add context into handler definition from the beginning? And there I think we could have better chances to use it for something useful, like proper context cancellation or tracing in the future.